ballista/rust/executor/src/executor_server.rs

```diff
@@ -289,18 +317,34 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> Heartbeater<T, U>
         Self { executor_server }
     }
 
-    async fn start(&self) {
+    fn start(&self, graceful: &Graceful) {
         let executor_server = self.executor_server.clone();
+        let mut heartbeat_shutdown = graceful.subscribe_for_shutdown();
+        // Not used directly.
+        let heartbeat_complete = graceful.shutdown_complete_tx.clone();
         tokio::spawn(async move {
             info!("Starting heartbeater to send heartbeat the scheduler periodically");
-            loop {
+            // Drop and notifies the receiver half that the shutdown is complete
```

Comment:
Maybe an explicit call to `drop` would be better in this case than a comment explaining that it will be dropped...

Reply:
Calling the drop explicitly is better in most case, but if the internal loop has complex branches, we will have to call the drop in all the possible branches which is error prone.

Reply:
That's a well-reasoned explanation, thank you :+1: